src/transformers/generation_logits_process.py

```diff
@@ -191,17 +191,14 @@ def __init__(self, top_p: float, filter_value: float = -float("Inf"), min_tokens
         self.min_tokens_to_keep = min_tokens_to_keep
 
     def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
-        sorted_logits, sorted_indices = torch.sort(scores, descending=True)
+        sorted_logits, sorted_indices = torch.sort(scores, descending=False)
         cumulative_probs = sorted_logits.softmax(dim=-1).cumsum(dim=-1)
 
         # Remove tokens with cumulative top_p above the threshold (token with 0 are kept)
-        sorted_indices_to_remove = cumulative_probs > self.top_p
+        sorted_indices_to_remove = cumulative_probs <= (1 - self.top_p)
```

Comment:
Isn't `<=` always a bit dangerous with float values? I'm not sure we can assure 100% backward compatibility here

Reply:
Slightly worried about that we'll silently break someone's PyTorch generation code that uses `top_p` by default here

Reply:
@patrickvonplaten there is indeed a change at the edge case -- before, if `top_p` was `0.8` and the input was `[0.5, 0.3, 0.1, 0.1]`, the first three tokens would pass this filter, despite the first two summing up to 0.8 (and thus satisfying the top P conditions, according to the original paper and our docstrings).

The behavior in TF and FLAX satisfies the edge case above, while PT does not. In practice, the impact will be negligible (this change filters one additional token when the sum of the logits is exactly `top_p`), although it can change seeded test cases.

Alternatively, we can change our docstrings (and TF+FLAX's implementation) to ignore this edge case :D

Reply:
@patrickvonplaten  I believe you are referring to the floating point precision in the context of `<=` being dangerous with float value. The Top P sampler intends to pick minimum elements which have cumulative dist >= `top_p`. So either we use the equality while selecting the mask or ignore it and then shift the mask to right/left. 

The proposed PT implementation uses `<=` but it can be implemented in the same manner as TF and FLAX which do not have the equality operator but will need to clone a tensor and shifting values to right/left. This however will not prevent the issue of floating point precision. 

E.g., if we take input as `[0.5, 0.3, 0.1, 0.1]` and `top_p` as `0.8` then according to this:
https://github.com/huggingface/transformers/blob/693ba2cc79ef46e8323b35f791233ce94f433b5d/src/transformers/generation_flax_logits_process.py#L142-L146

the cumsum _could_ be [0.5, 0.79995, 0.1, 0.1] due to floating point precision which will lead to Top P sampler picking 1st three elements instead of 1st two.

Reply:
Thanks for the explanations @gante and @ekagra-ranjan - this makes sense to me!

Given the very high usage of `generate` and `top_p` we need to clearly mark this as a "**breaking behavior bug fix**" with 🚨🚨🚨 in the PR description and also make sure it's mentioned in our release notes (cc @LysandreJik )

But good for merge then for me